cirq/ops/display.py

```diff
@@ -0,0 +1,206 @@
+        return self._pauli_string.qubits
+
+    def with_qubits(self,
+                    *new_qubits: raw_types.QubitId
+                    ) -> 'ApproxPauliStringExpectation':
+        return ApproxPauliStringExpectation(
+                self._pauli_string.with_qubits(*new_qubits),
+                self._num_samples,
+                self._key
+        )
+
+    @property
+    def key(self) -> Hashable:
+        return self._key
+
+    def measurement_basis_change(self) -> op_tree.OP_TREE:
+        return self._pauli_string.to_z_basis_ops()
+
+    @property
+    def num_samples(self) -> int:
+        return self._num_samples
+
+    def value_derived_from_samples(self,
+                                   measurements: np.ndarray) -> float:
+        return np.mean([(-1)**sum(bitstring) for bitstring in measurements])
+
+    def _value_equality_values_(self):
+        return self._pauli_string, self._num_samples, self._key
+
+
+@value.value_equality
+class PauliStringExpectation(WaveFunctionDisplay):
+    """Expectation value of a Pauli string."""
+
+    def __init__(self,
+                 pauli_string: 'pauli_string.PauliString',
+                 key: Hashable=''):
+        self._pauli_string = pauli_string
+        self._key = key
+
+    @property
+    def qubits(self) -> Tuple[raw_types.QubitId, ...]:
+        return self._pauli_string.qubits
+
+    def with_qubits(self,
+                    *new_qubits: raw_types.QubitId
+                    ) -> 'PauliStringExpectation':
+        return PauliStringExpectation(
+                self._pauli_string.with_qubits(*new_qubits),
+                self._key
+        )
+
+    @property
+    def key(self) -> Hashable:
+        return self._key
+
+    def value_derived_from_wavefunction(self,
+                                        state: np.ndarray,
+                                        qubit_map: Dict[raw_types.QubitId, int]
+                                        ) -> float:
+        num_qubits = state.shape[0].bit_length() - 1
+        ket = np.reshape(np.copy(state), (2,) * num_qubits)
+        for qubit, pauli in self._pauli_string.items():
+            buffer = np.empty(ket.shape, dtype=state.dtype)
+            args = protocols.ApplyUnitaryArgs(
+                    target_tensor=ket,
+                    available_buffer=buffer,
+                    axes=(qubit_map[qubit],)
+                    )
+            ket = protocols.apply_unitary(pauli, args)
+        ket = np.reshape(ket, state.shape)
+        return np.dot(state.conj(), ket)
+
+    def _value_equality_values_(self):
+        return self._pauli_string, self._key
+
+
+def pauli_string_expectation(
+        pauli_string: 'pauli_string.PauliString',
+        num_samples: Optional[int] = None,
+        key: Hashable='') -> Union[ApproxPauliStringExpectation,
+                                   PauliStringExpectation]:
+    if num_samples is None:
+        return PauliStringExpectation(pauli_string, key=key)
```

Comment:
I think we should go deeper and actually merge the two classes. This is going to be a really common thing that happens with displays.

Reply:
Oh, you mean have PauliStringExpectation inherit from both SamplesDisplay and WaveFunctionDisplay? That sounds good.

Reply:
Yes. It should be possible to do both at the same time.

Reply:
What should the abstract method `num_samples` of SamplesDisplay return if the user doesn't specify the number of samples? Seems like we need to choose an arbitrary default value. 1? 1000?

Reply:
Oh I just thought of one potential problem with having classes inherit from both SamplesDisplay and WaveFunctionDisplay. When running these on a simulator, how will the simulator know whether to compute the value of the display using the samples method or the wavefunction method?

Reply:
Or do you think `num_samples` of SamplesDisplay should be allowed to return None? I don't think that's a good idea.

Reply:
I think that this method should return something that isn't a samples display when given num_samples=None.

It would be ideal if we didn't need two distinct classes for the same concept, but if that's necessary behind the scenes then so be it. But I'd prefer that the user never realize that this is the case.

Reply:
So you're still insisting that the user instantiate Pauli string expectation displays by calling a class constructor instead of a normal method, right? Okay then, I will have to think about how to make that possible. As I mentioned above, inheritance doesn't seem like it will work here.

Reply:
Or, perhaps _all_ Displays should be able to have their values computed from the wavefunction. They could then opt into having a version that could be computed from samples.

Reply:
It's fine if they call a method. Preferable, even. It makes it easier to isolate them from changes to the objects behind the scene.

Reply:
That's what I have now ... what exactly is your complaint?

In any case, I'm thinking about renaming WaveFunctionDisplay to just Display and having SamplesDisplay be a subclass of it. The simulator would then always call the wavefunction method to compute the display value. This would let us merge the two PauliStringExpectation classes. What do you think of that?

Reply:
Actually, there would still be the question of what `num_samples` should return if the user inputs None.

Reply:
I don't think it makes sense for the sampling display to be a subclass of the wavefunction display. They're distinct input concepts.

I'm happy to have the two classes be separate and exposed via a common method. But it does seem like there's some kind of unifying concept we're missing here.

Reply:
My reasoning is that the value of any sampling display can in principle be computed from the wavefunction, so we could require that the wavefunction method always be present.

Reply:
I'm confused as to whether Displays are supposed to make it to the hardware or not.  The way this is written seems to suggest that you are only thinking about Display's for simulation?